Add a get method to retrieve entities from store without fetching for scenarios which rely on the entity being in the store.

mdgbayly opened the pull request:
We have a requirement in the rubrics editor to be able to utilize/access siren entities that are embedded as a child entity of a resource based entity, but which by themselves are not URI addressable - e.g. they don't have a self link. 

Could not come up with a great way to do this and to integrate it into our entity store. So I'm experimenting with the idea of flagging an entity with a `singleton` class. Entities which don't have an href or a self link href, but do have the singleton class will be added to the entity store keyed by their `rels`. So if you use a unique rel, then you can access the entity.

I've added a specific `get` store method that will not attempt to fetch an entity but just return it if it already exists in the store. This can be used to load these singleton entities.

All feels a bit hacky, but couldn't come up with a better way to access these non-resource based entities and we need something like that to support configuring the HTML editor in rubrics with information that comes from the LMS.

awikkerink commented:
Why not access the entities directly from the component instead of doing it in the store? I'm not understanding the problem. Do you have the code for rubric or a sample entity?

mdgbayly commented:
@awikkerink Here is the code that uses this:
https://github.com/Brightspace/d2l-rubric/pull/334/files#diff-ba99e695f7ccc4cb658ab755f3f9385bR167

Basically, we have rich text editor configuration information, that comes from the LMS, that we want to make available to each HTML editor component used to render the description for each rubric criterion cell and overall level.

So basically we have `application data` that is needed by deeply embedded sub components.

That application data is initially attached to the top level rubric entity, because we don't want to duplicate it on every deeply embedded entity (which was an option).

If we knew the URL of the top level rubric entity in those deeply embedded entities, we could just access it from the top level entity. But we don't, and I don't really want to have to pass the URL of the rubric entity down to these deeply embedded components either. So instead we store the data in the store using a well known fixed `rel` and the deeply embedded components can access the data directly using that `rel`.

We're just trying to avoid passing these `props` top-down (parent to child) as that can be cumbersome when they are required by many components in the application, particularly when the intermediate components have no use for them.

If we were using React for this app, then we'd probably use the [React Context](https://reactjs.org/docs/context.html) to make this available to sub-components.

If we were using Redux then we might just add it to the Redux store and allow the HTML editor components to pull it directly from the Redux store using a known store path.

awikkerink commented:
It feels like you're trying to use EntityStore like Redux. Where the url host/path represents a namespace.

I think this could work, but I don't really like it. If you would simply pass down the html editor config, how many components would it go through?

Alternatively, you could pass the `d2l-html-editor` in a slot?

mdgbayly commented:
@awikkerink Yeh you are right about using the url host/path as a namespace, and I agree it feels a bit hacky, but I think we are running up against limitations of trying to create our own `Store`. It works well for simple things where everything is a URI addressable resource, but as we're finding, that only gets you so far. And there are reasons why React and Redux provide features to work around the top-down props approach..

If we simply passed the html editor config down it would have to go through 7 levels of nesting.
For the same reason, I don't think it's feasible to pass the HTML editor in a slot. Not clear on how that would work anyway as we have multiple deeply nested HTML editors, and it seems like to pass it down from the top level component (if that is what you were suggesting) would make those top level components too knowledgeable about the lower level components?

The alternative I had considered was to make the editor config a URI addressable resource which could be included as a siren link in every rubric sub entity that had HTML editing requirements. And then loaded by those components using the standard fetch/listen approach. That way the store's existing de-duping would have resolved the duplication.

But this approach also felt unwieldy.  Standing up a new API endpoint to get the config information felt like overkill when it can be returned with the original rubric API call. We could just make it an additional rubric hypermedia API endpoint but that feels cludgy when it isn't really rubric specific.  But it also didn't seem like this data would neatly fit into any of our existing hypermedia domains, and standing up a new domain for `config` information also felt over the top.

Hence I decided to try the approach of making it easy to get at this kind of non-resource based `app data` from the store, which is most likely the approach I would have taken if using React/Redux.

awikkerink commented:
It's an interesting idea to put this logic in the EntityStore itself, but why not put it in using the `update` method in d2l-rubric?

mdgbayly commented:
@awikkerink 

Went with your suggestion of just having the rubric component add the editor config to the store using `update`.

So I removed the hacky, subversive `singleton` processing and just added a `get` method which can be used to get an entity from the store only if it already exists and doesn't initiate a fetch, which I think might be a generally useful function.

Now, the special handling of using the store to manage non-URI addressable entities is limited to how rubrics is using the store which feels less hacky and doesn't make the store take on any additional behaviors.
